src/subclass/application.rs

```diff
@@ -0,0 +1,530 @@
+// Copyright 2019, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use gio_sys;
+use glib_sys;
+
+use glib::translate::*;
+
+use glib::subclass::prelude::*;
+
+use Application;
+use ApplicationClass;
+
+use libc::{c_char, c_void};
+use std::convert;
+use std::ffi::OsString;
+use std::fmt;
+use std::ops::Deref;
+use std::ptr;
+
+pub struct ArgumentList {
+    pub(crate) ptr: *mut *mut *mut c_char,
+    items: Vec<OsString>,
+}
+
+impl ArgumentList {
+    pub(crate) fn new(arguments: *mut *mut *mut c_char) -> Self {
+        Self {
+            ptr: arguments,
+            items: unsafe { FromGlibPtrContainer::from_glib_none(ptr::read(arguments)) },
+        }
+    }
+
+    pub(crate) fn refresh(&mut self) {
+        self.items = unsafe { FromGlibPtrContainer::from_glib_none(ptr::read(self.ptr)) };
+    }
+
+    // remove the item at index `idx` and shift the raw array
+    pub fn remove(&mut self, idx: usize) {
+        unsafe {
+            let n_args = glib_sys::g_strv_length(*self.ptr);
+            assert!((n_args as usize) == self.items.len());
+            assert!((idx as u32) < n_args);
+
+            self.items.remove(idx);
+
+            glib_sys::g_free(((*self.ptr).offset(idx as isize)) as *mut c_void);
+
+            for i in (idx as u32)..n_args - 1 {
+                ptr::write(
+                    (*self.ptr).offset(i as isize),
+                    *(*self.ptr).offset((i + 1) as isize),
+                )
+            }
+            ptr::write((*self.ptr).offset((n_args - 1) as isize), ptr::null_mut());
+        }
+    }
+}
+
+impl Deref for ArgumentList {
+    type Target = [OsString];
+
+    fn deref(&self) -> &Self::Target {
+        self.items.as_slice()
+    }
+}
+
+impl fmt::Debug for ArgumentList {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        self.items.fmt(formatter)
+    }
+}
+
+impl convert::Into<Vec<OsString>> for ArgumentList {
+    fn into(self) -> Vec<OsString> {
+        self.items.clone()
```

Comment:
Clone really needed? Why not replace it with empty vector?

Reply:
Indeed it should be possible to just move it out

Reply:
Fixed. Thanks :)